pkg/urbit/worker/mars.c

```diff
@@ -110,64 +118,57 @@ _mars_grab(u3_noun sac)
 }
 
 /* _mars_fact(): commit a fact and enqueue its effects.
-*/
+ */
 static void
-_mars_fact(u3_mars* mar_u,
-           u3_noun    job,
-           u3_noun    pro)
+_mars_fact(u3_mars* mar_u, u3_noun job, u3_noun pro)
 {
   {
-    u3_fact tac_u = {
-      .job   = job,
-      .mug_l = mar_u->mug_l,
-      .eve_d = mar_u->dun_d
+    u3_gift gif_u = {
+      .sat_e = u3_gift_fact_e,
+      .eve_d = mar_u->dun_d,
     };
 
-    u3_disk_plan(mar_u->log_u, &tac_u);
-    u3z(job);
+    u3s_jam_xeno(pro, &gif_u.len_d, &gif_u.hun_y);
+    u3z(pro);
+
+    c3_list_pushb(mar_u->gif_u, &gif_u, sizeof(gif_u));
   }
 
   {
-    u3_gift* gif_u = c3_malloc(sizeof(*gif_u));
-    gif_u->nex_u = 0;
-    gif_u->sat_e = u3_gift_fact_e;
-    gif_u->eve_d = mar_u->dun_d;
+    u3_atom mat = u3ke_jam(job);
+    c3_w  len_w = u3r_met(3, mat);
 
-    u3s_jam_xeno(pro, &gif_u->len_d, &gif_u->hun_y);
-    u3z(pro);
+    c3_l   mug_l = mar_u->mug_l;
+    size_t len_i = sizeof(mug_l) + len_w;
+    c3_y   dat_y[len_i];
+    dat_y[0] = mug_l & 0xff;
+    dat_y[1] = (mug_l >> 8) & 0xff;
+    dat_y[2] = (mug_l >> 16) & 0xff;
+    dat_y[3] = (mug_l >> 24) & 0xff;
+    u3r_bytes(0, len_w, dat_y + sizeof(mug_l), mat);
+    u3z(mat);
 
-    if ( !mar_u->gif_u.ent_u ) {
-      c3_assert( !mar_u->gif_u.ext_u );
-      mar_u->gif_u.ent_u = mar_u->gif_u.ext_u = gif_u;
-    }
-    else {
-      mar_u->gif_u.ent_u->nex_u = gif_u;
-      mar_u->gif_u.ent_u = gif_u;
-    }
+    u3_saga_commit_async(mar_u->log_u,
+                         dat_y,
+                         len_i);
   }
+
```

Comment:
This should go back to commit first, then queue effects. If the event log is idle, that async commit can begin earlier that way.